### PR TITLE
[SAI-PTF]Add SAI-PTF Warm reboot design doc

### PR DESCRIPTION
**Pull request by richardyu-ms**
Add design doc for the SAI-PTF warm reboot.

It contains the introduction about the warm reboot in SAI-PTF, and
provide the sample code, command line.

Signed-off-by: richardyu-ms <richard.yu@microsoft.com>

### CODE REVIEW

**Comment by rlhui** [comment]
Can we add a network topology diagram for the warmboot tests

**Comment by rlhui** [comment]
Please add test cases details for the warmboot SAI tests

File: doc/sai-ptf/warmboot.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+```
+@warm_setup
+@warm_test
+@warm_teardown
+```
+there is the sample code
+- Setup methods
+   ```python
+    @warm_setup
+    def setUp(self):
```
Comment (ravi861) [inline doc/sai-ptf/warmboot.md:67]: Can you please list which of these 9 stages execute together at each of the 3 stages?

Comment (richardyu-ms) [reply doc/sai-ptf/warmboot.md:67]: do you mean this list?

- Before start, three methods
  - setup, runTest, tearDown 
- Starting, three methods
  - setUp_starting, test_starting, tearDown_starting
- After Starting, three methods.
  - setUp_post_start, test_post_start, tearDown_post_start

**Comment by richardyu-ms** [comment]
> Can we add a network topology diagram for the warmboot tests

diagram added

**Comment by richardyu-ms** [comment]
> Please add test cases details for the warmboot SAI tests

Description added and describe the process for creating a warm reboot test

File: doc/sai-ptf/warmboot.md [hunk 2]
```diff
@@ -2,36 +2,41 @@
...
+For SAI-PTF, it will use a non-topology network structure for the sai testing. 
+
+![Device_topology](img/SAI_PTF_Topology.jpg)
+### Testbed
+Those tests will be run on the testbed structure as below, the components are:
+* PTF - running in a server that can connect to the target DUT
 * SAI server - running on a dut
    ![Device_topology](img/Device_topology.jpg)
-*p.s. cause the sai testing will not depends on any sonic components, then there will be no specifical topology(T0 T1 T2) for testing.*
+*p.s. cause theSAtesting will not depend on any sonic components, then there will be no specific topology(T0 T1 T2) for testing.*
```
Comment (aljer) [inline doc/sai-ptf/warmboot.md:33]: typo and missing spaces

File: doc/sai-ptf/warmboot.md [hunk 3]
```diff
@@ -215,7 +220,41 @@ The code for making the warm shutdown is
...
+  1. create Vlans for each port
+  2. create MACs for ports (mapped with index)
+  3. then create route rules by adding mac and port
+  4. then send packets to on port by with different mac
+- Check
+
+  5. check we can get the port allocated with the mac as route rules
+
+With that basic VLAN and FDB functionality, we expect in a warm reboot scenario, we can get the test case and expectations below
+1. pre-warmreboot
```
Comment (aljer) [inline doc/sai-ptf/warmboot.md:248]: minor: start with capital letter

File: doc/sai-ptf/warmboot.md [hunk 4]
```diff
@@ -215,7 +220,41 @@ The code for making the warm shutdown is
...
+
+In order to reuse the existing test cases in a warm reboot, we need to re-organize the existing test cases into the warm reboot structure.
+
+### Test case detail
+There is a test sample for the warm reboot test https://github.com/opencomputeproject/SAI/pull/1440
+
+
+For this warm reboot test in the sample, we reuse the code in the [L2SanityTest](https://github.com/richardyu-ms/SAI/blob/v1.7/ptf/saisanity.py). In general, in this test case, we 
+- Setup
+  1. create Vlans for each port
```
Comment (aljer) [inline doc/sai-ptf/warmboot.md:239]: VLANs

File: doc/sai-ptf/warmboot.md [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+   sai_warmboot.sh
+   ```
+1. change the profile, enable the warm start
+   ```shell
+   # Sample in a dell s6000
+   cat /etc/sai.d/sai.profile
+   SAI_INIT_CONFIG_FILE=/usr/share/sonic/hwsku/td2-s6000-32x40G.config.bcm
+   SAI_NUM_ECMP_MEMBERS=32
+   SAI_WARM_BOOT_WRITE_FILE=/var/warmboot/sai-warmboot.bin
+   SAI_WARM_BOOT_READ_FILE=/var/warmboot/sai-warmboot.bin
```
Comment (rlhui) [inline doc/sai-ptf/warmboot.md:205]: Can we add what does the warm reboot SAI test do with these two files

File: doc/sai-ptf/warmboot.md [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+        return f(inst)
+    return test_director
+```
+
+## Support DUT different running stages during warm reboot
+During a warm reboot, we can split the DUT into three different stage
+- Before starting, in this stage, we will make some configurations or benchmark on the DUT, and use those setting to check the DUT functionality during a warm reboot. All the preparation for the warm reboot will be finished.
+- Starting, in this stage, DUT will run a warm-reboot command and not start any other docker services and sai-sdk will not start. We expect all the traffic will behave as normal.
+- After Starting, in this stage, sai-sdk will start, and data will be restored from the backup file. We expect there should be no change on the route and all the traffic will behave as normal. Meanwhile, if needed we can check the switch status from some SAI interfaces, as in this stage, SAI switch has been initialized and RPC APIs have been ready.
+
```
Comment (rlhui) [inline doc/sai-ptf/warmboot.md:148]: It's better to move the above section closer to line 54 ("During a warm reboot, we need to split those three method in PTF-Test into three different stage...)

File: doc/sai-ptf/warmboot.md [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+2. Starting
+    - Check as a test in L2SanityTest
+
+*In this stage, like a switch running in starting mode, sai switch is not getting started, we will only verify the packet transit*
+    
+3. Post-warm-reboot
+    - APIs start the switch in warm mode
+    - Check as a test in L2SanityTest
+
+*In this stage, the switch gets started, and we can use warm reboot API to start the switch.*
```
Comment (rlhui) [inline doc/sai-ptf/warmboot.md:262]: Can we elaborate in the doc what exactly is being verified after warmboot?  Also can we add reconciliation /ensuring that SW entries are same as HW entries (and not just by using packets)?

**Comment by richardyu-ms** [comment]
> Please add test cases details for the warmboot SAI tests

added the comments in this commit 
https://github.com/opencomputeproject/SAI/pull/1449/commits/24c9f02375d7d93f3186eca0ec8ab18c06465eee

According to the review meeting, I will add more for this part.

**Review by ravi861 — APPROVED** [review]

File: doc/sai-ptf/warmboot.md (hunk 6 above)
Comment (richardyu-ms) [reply doc/sai-ptf/warmboot.md:148]: ok, move this section upper to the section to introduce the different warm reboot stages

File: doc/sai-ptf/warmboot.md (hunk 7 above)
Comment (richardyu-ms) [reply doc/sai-ptf/warmboot.md:293]: added a conclusion in the end

**Pull request closed without merging**